### PR TITLE
[Feat] Add LabelStudio ML Backend

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds an ImageClassifierAPI to be used for ActiveLearning directly with LabelStudio.

Fixes # (issue)

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/image/classification/integrations/labelstudio/handler.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            image_urls.append(completion["data"][self.value])
+            image_classes.append(get_choice(completion))
+
+        train_files = [self._download_image(url, url_class) for url, url_class in zip(image_urls, image_classes)]
+        class_to_id = {c: id for id, c in enumerate(np.unique(image_classes))}
+        train_targets = np.vectorize(class_to_id.get)(image_classes)
+        datamodule = ImageClassificationData.from_files(train_files=train_files, train_targets=train_targets)
+        trainer = flash.Trainer(max_epochs=1, gpus=int(torch.cuda.is_available()))
+        trainer.fit(self.model, datamodule=datamodule)
+        model_path = os.path.join(".", "model.pt")
```
Comment (niklub) [inline flash/image/classification/integrations/labelstudio/handler.py:111]: there is a `workdir` folder that serves for storing checkpoints from the current run. Unless you want to overwrite the same model.pt file each time, I would recommend to use `os.path.join(workdir, "model.pt")`

Comment (tchaton) [reply flash/image/classification/integrations/labelstudio/handler.py]: Done !

File: flash/image/classification/integrations/baal/dropout.py [hunk 2]
```diff
@@ -31,10 +33,24 @@ def __init__(self, module: flash.Task, inference_iteration: int):
...
         for _ in range(self.inference_iteration):
             out.append(self.parent_module.predict_step(batch, batch_idx, dataloader_idx=dataloader_idx))
 
         # BaaL expects a shape [num_samples, num_classes, num_iterations]
         return torch.tensor(out).permute((1, 2, 0))
+
+    def on_predict_dataloader(self) -> None:
+        if self.parent_module.data_pipeline is not None:
+            self.parent_module.data_pipeline._detach_from_model(self.parent_module, RunningStage.PREDICTING)
+            self.parent_module.data_pipeline._attach_to_model(self.parent_module, RunningStage.PREDICTING)
```
Comment (SeanNaren) [inline flash/image/classification/integrations/baal/dropout.py:55]: Could you explain what's happening here?

File: flash/image/classification/integrations/labelstudio/handler.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+        self._IMAGE_CACHE_DIR = os.path.join(os.path.dirname(source), "image_cache")
+        os.makedirs(self._IMAGE_CACHE_DIR, exist_ok=True)
+
+    def predict(self, tasks, **kwargs):
+        image_urls = [self._firstv(task["data"]) for task in tasks]
+        predict_files = [self._download_image(url) for url in image_urls]
+
+        datamodule = ImageClassificationData.from_files(predict_files=predict_files, batch_size=2)
+        trainer = flash.Trainer(max_epochs=1, gpus=int(torch.cuda.is_available()))
```
Comment (SeanNaren) [inline flash/image/classification/integrations/labelstudio/handler.py:80]: This is pretty hardcoded, any way this can be exposed?

Comment (tchaton) [reply flash/image/classification/integrations/labelstudio/handler.py:80]: Not yet.

File: flash/image/classification/integrations/baal/dropout.py (hunk 2 above)
Comment (tchaton) [reply flash/image/classification/integrations/baal/dropout.py:55]: The predict_step is being automatically wrap with the postprocessor and serializer. Without overriding the function, it would wrap the InferenceModel and not its children.

Comment (tchaton) [reply flash/image/classification/integrations/baal/dropout.py:55]: This enable self.parent_model.predict_step to actually returns the Probabilities.

File: flash/image/classification/integrations/labelstudio/handler.py (hunk 3 above)
Comment (tchaton) [reply flash/image/classification/integrations/labelstudio/handler.py:80]: But there might be some solutions.

File: flash/image/classification/integrations/baal/dropout.py (hunk 2 above)
Comment (SeanNaren) [reply flash/image/classification/integrations/baal/dropout.py:55]: Could you explain in a comment in the code @tchaton? it looks as if you're detaching the predict pipeline then re-attaching it straight after with no changes whatsoever.

File: flash/image/classification/integrations/labelstudio/handler.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            if is_skipped(completion):
+                continue
+            image_urls.append(completion["data"][self.value])
+            image_classes.append(get_choice(completion))
+
+        train_files = [self._download_image(url, url_class) for url, url_class in zip(image_urls, image_classes)]
+        class_to_id = {c: id for id, c in enumerate(np.unique(image_classes))}
+        train_targets = np.vectorize(class_to_id.get)(image_classes)
+        datamodule = ImageClassificationData.from_files(train_files=train_files, train_targets=train_targets)
+        trainer = flash.Trainer(max_epochs=1, gpus=int(torch.cuda.is_available()))
```
Comment (SeanNaren) [inline flash/image/classification/integrations/labelstudio/handler.py:119]:
```suggestion
        # currently label studio does not support passing parameters to the learning loop.
        trainer = flash.Trainer(max_epochs=1, gpus=int(torch.cuda.is_available()))
```

File: setup.cfg [hunk 5]
```diff
@@ -22,6 +22,9 @@ exclude_lines =
     if __name__ == .__main__.:
     add_model_specific_args
 
+[coverage:run]
+omit =
+    flash/image/classification/integrations/labelstudio/handler.py
```
Comment (SeanNaren) [inline setup.cfg:27]: you explained something around this being it took too long for the tests to run, is this still the case or have I misunderstood @tchaton?

Comment (tchaton) [reply setup.cfg:27]: I didn't check it out. I need to test it out.
